src/JustEat.StatsD/StatsDPublisher.cs

```diff
@@ -1,146 +1,156 @@
 using System;
+using JustEat.StatsD.Buffered;
 using JustEat.StatsD.EndpointLookups;
 
 namespace JustEat.StatsD
 {
+    // ReSharper disable IntroduceOptionalParameters.Global to preserve binary compatibility
+
     /// <summary>
     ///     Will synchronously publish stats at statsd as you make calls; will not batch sends.
     /// </summary>
     public class StatsDPublisher : IStatsDPublisher
     {
-        private readonly StatsDMessageFormatter _formatter;
-        private readonly IStatsDTransport _transport;
-        private readonly Func<Exception, bool> _onError;
+        private readonly IStatsDPublisher _publisher;
 
         public StatsDPublisher(StatsDConfiguration configuration, IStatsDTransport transport)
+            : this (configuration, transport, true)
+        {
+        }
+
+        public StatsDPublisher(StatsDConfiguration configuration, IStatsDTransport transport, bool preferBufferedTransport)
         {
             if (configuration == null)
             {
-               throw new ArgumentNullException(nameof(configuration));
+                throw new ArgumentNullException(nameof(configuration));
             }
 
-            _transport = transport ?? throw new ArgumentNullException(nameof(transport));
+            if (transport == null)
+            {
+                throw new ArgumentNullException(nameof(transport));
+            }
 
-            _formatter = new StatsDMessageFormatter(configuration.Prefix);
-            _onError = configuration.OnError;
+            switch (transport)
+            {
+                case IStatsDBufferedTransport transportV2 when preferBufferedTransport:
+                    _publisher = new BufferBasedStatsDPublisher(configuration, transportV2);
+                    break;
+                default:
+                    _publisher = new StringBasedStatsDPublisher(configuration, transport);
+                    break;
+            }
         }
 
         public StatsDPublisher(StatsDConfiguration configuration)
+            : this(configuration, true)
+        {
+        }
+
+        public StatsDPublisher(StatsDConfiguration configuration, bool preferBufferedTransport)
         {
             if (configuration == null)
             {
-               throw new ArgumentNullException(nameof(configuration));
+                throw new ArgumentNullException(nameof(configuration));
             }
 
-            if (string.IsNullOrWhiteSpace(configuration.Host))
+            var endpointSource = EndpointParser.MakeEndPointSource(
+                configuration.Host, configuration.Port, configuration.DnsLookupInterval);
+
+            var transport = new PooledUdpTransport(endpointSource);
+
+            if (preferBufferedTransport)
```

Comment:
As alluded to in my comment, maybe we could have `if (configuration.PreferBufferedTransport)` instead?

Reply:
Now it seems so obvious, thx @martincostello